pkg/server/skydb/access.go

```diff
@@ -126,3 +130,192 @@ func (acl RecordACL) Accessible(userinfo *UserInfo, level RecordACLLevel) bool {
+	acl := NewFieldACL(list)
+	entry := acl.FindDefaultEntry()
+	if entry == nil {
+		// There is no entry with wildcard record type and record field,
+		// add the default entry to the wildcardRecordType list
+		defEntry.RecordType = WildcardRecordType
+		defEntry.RecordField = WildcardRecordField
+		defEntry.UserRole = "_public"
+		acl.wildcardRecordType = append(acl.wildcardRecordType, defEntry)
+	}
+
+	return acl
+}
+
+// AllEntries return all ACL entries in FieldACL.
+func (acl FieldACL) AllEntries() FieldACLEntryList {
+	result := acl.wildcardRecordType
+	for _, entries := range acl.recordTypes {
+		result = append(result, entries...)
+	}
+	return result
+}
+
+// FindDefaultEntry finds the default ACL entry in FieldACL.
+//
+// This function returns nil if the default ACL entry is not contained
+// in the FieldACL.
+func (acl FieldACL) FindDefaultEntry() *FieldACLEntry {
+	return acl.wildcardRecordType.findDefaultEntry()
+}
+
+// Accessible returns true when the access mode is allowed access
+func (acl FieldACL) Accessible(
+	userinfo *UserInfo,
+	record *Record,
+	recordType string,
+	field string,
+	mode FieldAccessMode,
+) bool {
+	if acl.wildcardRecordType.Accessible(userinfo, record, recordType, field, mode) {
+		return true
+	}
+	if list, ok := acl.recordTypes[recordType]; ok {
+		return list.Accessible(userinfo, record, recordType, field, mode)
+	}
+	return false
+}
+
+// FieldACLEntryList contains a list of field ACL entries
+type FieldACLEntryList []FieldACLEntry
+
+// Accessible returns true when the access mode is allowed access
+func (list FieldACLEntryList) Accessible(
+	userinfo *UserInfo,
+	record *Record,
+	recordType string,
+	field string,
+	mode FieldAccessMode,
+) bool {
+	for _, entry := range list {
+		if entry.Accessible(userinfo, record, recordType, field, mode) {
+			return true
+		}
+	}
+	return false
+}
+
+func (list FieldACLEntryList) findDefaultEntry() *FieldACLEntry {
+	for _, entry := range list {
+		if entry.RecordType == WildcardRecordType &&
+			entry.RecordField == WildcardRecordField &&
+			entry.UserRole == "_public" {
+			return &entry
+		}
+	}
+	return nil
+}
+
+func (list FieldACLEntryList) Len() int      { return len(list) }
+func (list FieldACLEntryList) Swap(i, j int) { list[i], list[j] = list[j], list[i] }
+func (list FieldACLEntryList) Less(i, j int) bool {
+	// compare is similar to strings.Compare except that specified wildcard
+	// string will be less than non-wildcard string.
+	compare := func(a, b, wildcard string) int {
+		if a == wildcard && b != wildcard {
+			return -1
+		} else if b == wildcard && a != wildcard {
+			return 1
+		}
+		return strings.Compare(a, b)
+	}
+
+	result := compare(list[i].RecordType, list[j].RecordType, WildcardRecordType)
+	if result != 0 {
+		return result < 0
+	}
+
+	result = compare(list[i].RecordField, list[j].RecordField, WildcardRecordField)
+	if result != 0 {
+		return result < 0
+	}
+
+	return strings.Compare(list[i].UserRole, list[j].UserRole) < 0
```

Comment:
I am working on a branch that uses struct in `UserRole`. So I will mark this as TODO.